[4/N] Add backend options map

cccclai opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #11758
* __->__ #11462
* #11391
* #11390
* #11389



This is to manage the backend <-> BackendOptions map. Users will create the bakcend options map, and ET runtime will read the backend name, and dispatch the list of backend options to each backend.
@exported-using-ghexport

Differential Revision: [D76149466](https://our.internmc.facebook.com/intern/diff/D76149466/)

Differential Revision: [D76149466](https://our.internmc.facebook.com/intern/diff/D76149466)

cccclai committed:
[4/N] Add backend options map

This is to manage the backend <-> BackendOptions map. Users will create the bakcend options map, and ET runtime will read the backend name, and dispatch the list of backend options to each backend.

Differential Revision: [D76149466](https://our.internmc.facebook.com/intern/diff/D76149466/)

[ghstack-poisoned]

cccclai committed:
Update on "[4/N] Add backend options map"


This is to manage the backend <-> BackendOptions map. Users will create the bakcend options map, and ET runtime will read the backend name, and dispatch the list of backend options to each backend.

Differential Revision: [D76149466](https://our.internmc.facebook.com/intern/diff/D76149466/)

[ghstack-poisoned]

cccclai committed:
Update on "[4/N] Add backend options map"


This is to manage the backend <-> BackendOptions map. Users will create the bakcend options map, and ET runtime will read the backend name, and dispatch the list of backend options to each backend.

Differential Revision: [D76149466](https://our.internmc.facebook.com/intern/diff/D76149466/)

[ghstack-poisoned]

kimishpatel commented on runtime/backend/backend_options.h:
```diff
@@ -6,94 +6,92 @@
- } // namespace executorch
- 
\ No newline at end of file
+#pragma once
```
I am not sure whats happening in these diffs and why they show up as so many deleted lines and so many added lines

cccclai replied:
Ah it's the weird formatting. Let me update so it's easier to review. I've been manually running lintrunner to get rid of the weird formating

cccclai replied:
Changes in backend_options.h is not needed...clean up the PR here

cccclai committed:
Update on "[4/N] Add backend options map"


This is to manage the backend <-> BackendOptions map. Users will create the bakcend options map, and ET runtime will read the backend name, and dispatch the list of backend options to each backend.

Differential Revision: [D76149466](https://our.internmc.facebook.com/intern/diff/D76149466/)

[ghstack-poisoned]

cccclai committed:
Update on "[4/N] Add backend options map"


This is to manage the backend <-> BackendOptions map. Users will create the bakcend options map, and ET runtime will read the backend name, and dispatch the list of backend options to each backend.

Differential Revision: [D76149466](https://our.internmc.facebook.com/intern/diff/D76149466/)

[ghstack-poisoned]

cccclai committed:
Update on "[4/N] Add backend options map"




This is to manage the backend <-> BackendOptions map. Users will create the bakcend options map, and ET runtime will read the backend name, and dispatch the list of backend options to each backend.
exported-using-ghexport

Differential Revision: [D76149466](https://our.internmc.facebook.com/intern/diff/D76149466/)

Differential Revision: [D76149466](https://our.internmc.facebook.com/intern/diff/D76149466)

[ghstack-poisoned]

cccclai committed:
Update on "[4/N] Add backend options map"




This is to manage the backend <-> BackendOptions map. Users will create the bakcend options map, and ET runtime will read the backend name, and dispatch the list of backend options to each backend.
exported-using-ghexport

Differential Revision: [D76149466](https://our.internmc.facebook.com/intern/diff/D76149466/)

Differential Revision: [D76149466](https://our.internmc.facebook.com/intern/diff/D76149466)

[ghstack-poisoned]

cccclai committed:
Update on "[4/N] Add backend options map"




This is to manage the backend <-> BackendOptions map. Users will create the bakcend options map, and ET runtime will read the backend name, and dispatch the list of backend options to each backend.
exported-using-ghexport

Differential Revision: [D76149466](https://our.internmc.facebook.com/intern/diff/D76149466/)

Differential Revision: [D76149466](https://our.internmc.facebook.com/intern/diff/D76149466)

[ghstack-poisoned]

cccclai committed:
Update on "[4/N] Add backend options map"




This is to manage the backend <-> BackendOptions map. Users will create the bakcend options map, and ET runtime will read the backend name, and dispatch the list of backend options to each backend.
exported-using-ghexport

Differential Revision: [D76149466](https://our.internmc.facebook.com/intern/diff/D76149466/)

Differential Revision: [D76149466](https://our.internmc.facebook.com/intern/diff/D76149466)

[ghstack-poisoned]

cccclai committed:
Update on "[4/N] Add backend options map"




This is to manage the backend <-> BackendOptions map. Users will create the bakcend options map, and ET runtime will read the backend name, and dispatch the list of backend options to each backend.
exported-using-ghexport

Differential Revision: [D76149466](https://our.internmc.facebook.com/intern/diff/D76149466/)

Differential Revision: [D76149466](https://our.internmc.facebook.com/intern/diff/D76149466)

[ghstack-poisoned]

kimishpatel commented on runtime/backend/backend_options_map.h:
```diff
@@ -0,0 +1,74 @@
+};
+
+template <size_t MaxBackends>
+class BackendOptionsMap {
```
Can you not just use set_option and get_option? We discussed this by allowing users to directly set backend options, without having to plumb it through methods/program etc., and I belive you did that? In that case do we still need this?

cccclai replied:
backend options map is for setting multiple backends, users code will be like following
```
// users' code
BackendOptions backend_options = {
    {"CoreMLBackend", {{"power", "high"}}},
    {"QNNBackend", {{"backend", "CPU_AND_ANE"}}}
};

// Optional
update(backend_options);

module.load(file);
// Optional
update(backend_options);
    
module.execute(inputs);

```

cccclai committed:
Update on "[4/N] Add backend options map"




This is to manage the backend <-> BackendOptions map. Users will create the bakcend options map, and ET runtime will read the backend name, and dispatch the list of backend options to each backend.
exported-using-ghexport

Differential Revision: [D76149466](https://our.internmc.facebook.com/intern/diff/D76149466/)

Differential Revision: [D76149466](https://our.internmc.facebook.com/intern/diff/D76149466)

[ghstack-poisoned]

kimishpatel reviewed (approved):
I am not 100% convinced we need this, but not strongly opposed to it. Leave it to the author
